lib/semantics/resolve-labels.cc

```diff
@@ -311,6 +320,156 @@ class ParseTreeAnalyzer {
     popConstructNameWithoutBlock(forallConstruct);
   }
 
+  // C1414
+  void Post(const parser::BlockData &blockData) {
+    if (!BothEqOrNone(
+            std::get<parser::Statement<parser::BlockDataStmt>>(blockData.t)
+                .statement.v,
+            std::get<parser::Statement<parser::EndBlockDataStmt>>(blockData.t)
+                .statement.v)) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{"BLOCK DATA name mismatch"_err_en_US});
```

Comment:
In these messages it would be friendly to include a pointer to the first name. It makes it easier for the user to spot the mismatch.